calico/felix/fiptables.py

```diff
@@ -1,214 +1,451 @@
+IPTABLES_CMD = { IPV4: "iptables",
+                 IPV6: "ip6tables" }
+
+# Logger
+log = logging.getLogger(__name__)
+
+# Special value to mean "put this rule at the end".
+RULE_POSN_LAST = -1
+
+class Chain(object):
+    """
+    Representation of an iptables chain.
+    """
+    def __init__(self, table, name):
+        self.type = table.type
+        self.table = table
+        self.name = name
+        self.rules = []
+        self.table.chains[name] = self
+
+    def flush(self):
+        log.debug("Flushing chain %s", self.name)
+        self.table.ops.append([IPTABLES_CMD[self.type],
+                               "-w",
+                               "-t",
+                               self.table.name,
+                               "-F",
+                               self.name])
+        del self.rules[:]
+
+    def delete_rule(self, rule):
+        # The rule must exist or it is an error.
+        self.rules.remove(rule)
+        args = [IPTABLES_CMD[self.type],
+                "-w",
+                "-t",
+                self.table.name,
+                "-D",
+                self.name]
+        args.extend(rule.generate_fields())
+        self.table.ops.append(args)
+        log.debug("Removing rule : %s", args)
+
+    def truncate_rules(self, count):
+        """
+        This is a utility function to remove any excess rules from a
+        chain. After we have carefully inserted all the rules we want at the
+        start, we want to get rid of any legacy rules from the end.
+
+        It takes a chain object, and a count for how many of the rules should
+        be left in place.
+        """
+        log.debug("Truncate %s chain %s to length %d (length now %d)",
+                  self.type,
+                  self.name,
+                  count,
+                  len(self.rules))
+
+        while len(self.rules) > count:
+            self.table.ops.append([IPTABLES_CMD[self.type],
+                                   "-w",
+                                   "-t",
+                                   self.table.name,
+                                   "-D",
+                                   self.name,
+                                   str(count + 1)])
+            self.rules.pop(count)
+
+
+    def insert_rule(self, rule, position=0, force_position=True):
+        """
+        Add an iptables rule to a chain if it does not already exist. Position
+        is the position for the insert as an offset; if set to RULE_POSN_LAST
+        then the rule is appended.
+
+        If force_position is True, then the rule is added at the specified point
+        unless it already exists there. If force_position is False, then the
+        rule is added only if it does not exist anywhere in the list of rules.
+
+        Position starts at offset 0, even though iptables wants it to start at
+        offset 1.
+        """
+        if position == RULE_POSN_LAST:
+            position = len(self.rules)
```

Comment:
We have to wait for the lock because somebody (such as the diags script) might be reading the data under our feet. If somebody is modifying any of our chains under our feet, we are pretty much sunk.

The only case we are left with is where we are appending a rule to a chain that we do not own, where it only really matters that it gets in there, not where, so I think I agree that we might as well use append in that case.